### PR TITLE
Ericsnow resource collection

### PR DESCRIPTION
**Pull request by mhilton**
Make review updates to the branch in #580

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1699/
Test PASSed.

File: internal/charmstore/resource.go [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"sort"
+
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/mgo.v2/bson"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+)
+
+// newResourceQuery formats the provided details into a mongo query for
```
Comment (rogpeppe) [inline internal/charmstore/resource.go]: // newResourceQuery returns a mongo query doc that will retrieve the
// given named resource and revision associated with the given
// charm URL. If revision is < 0, all revisions of the resource
// will be selected by the query.

?

File: internal/charmstore/resource.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	if revision >= 0 {
+		query = append(query, bson.DocElem{"revision", revision})
+	}
+	return query
+}
+
+// sortResources sorts the provided resource docs, The resources are
+// sorted first by URL then by name and finally by revision.
+func sortResources(resources []*mongodoc.Resource) {
+	sort.Sort(resourcesByName(resources))
```
Comment (rogpeppe) [inline internal/charmstore/resource.go:31]: ... or just inline where needed, I guess.

File: internal/charmstore/resource_test.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	expectResources: []*mongodoc.Resource{{
+		BaseURL:  charm.MustParseURL("~bob/wordpress"),
+		Name:     "res",
+		Revision: 1,
+	}},
+}}
+
+func (s *resourceSuite) TestNewResourceQuery(c *gc.C) {
+	for _, r := range []string{
+		"~bob/wordpress|res|0",
```
Comment (rogpeppe) [inline internal/charmstore/resource_test.go:101]: We might consider standardising on a syntax for this I guess
and making a helper function.

File: internal/mongodoc/resource_test.go [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+	jc "github.com/juju/testing/checkers"
+	gc "gopkg.in/check.v1"
+	"gopkg.in/juju/charm.v6-unstable"
+	"gopkg.in/mgo.v2/bson"
+
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+)
+
+const fingerprint = "0123456789abcdef0123456789abcdef0123456789abcdef"
```
Comment (rogpeppe) [inline internal/mongodoc/resource_test.go]: fakeBlobHash ?

**Comment by rogpeppe** [comment]
LGTM with a few trivial suggestions, thanks!

File: internal/charmstore/resource.go (hunk 2 above)
Comment (mhilton) [inline internal/charmstore/resource.go:31]: I'll leave this for now, it seems to be a useful enough function.

File: internal/charmstore/resource_test.go (hunk 3 above)
Comment (mhilton) [inline internal/charmstore/resource_test.go:101]: Good idea, I'll see what tests we do in later PRs and work with this appropriately.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/1700/
Test FAILed.

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**